Camera AV Stream Management Server Impl

pidarped opened the pull request:
* Delegate definition
  - Support for Command handlers
  - Helper interfaces for loading of pre-allocated streams from storage

* Attributes support
  - Read/Write support
  - Setters/Getters for all attributes
  - Storage handling for most attributes, except streams.
  
* Add attributes to json files

* Autogen code changes to zzz_generated/

#### Testing

pidarped committed:
Camera AV Stream Management Server Impl

* Delegate definition
  - Support for list entries per index
  - Support for Command handlers

* Server Impl
  - Attributes and Commands support
    -- Setters for Application to update values

* Add attributes to json files

* Autogen code changes to zzz_generated/

pidarped committed:
Add Optional Attributes support

pidarped committed:
Add support logic for writable attributes.

Modify feature checks for the attributes depending on their optionality
and feature categories.

pidarped committed:
Remove some attributes to simplify AVStreamMgmt cluster slightly.

Attributes related to White balance, ShutterSPeed, ISO and Depth sensor
were removed based on feedback at the Nov member meeting.
Also, adjustments for read/write mode changes for certain attributes.

pidarped committed:
Remove redundant CurrentSnapshotConfig and CurrentVideoCodecs attributes

Fix name of MaxFramRate in Snapshot Allocation functions.

pidarped committed:
Add write support for ImageRotation, ImageFlipHorizontal and ImageFlipVertical attributes.

pidarped committed:
Add handler for the SetStreamPriorities command.

pidarped committed:
Retrieve persistent attributes from storage during initialization.

pidarped committed:
Miscellaneous syntactic fixes to CameraAVStreamMgmt.

pidarped committed:
Add persistence for Viewport struct attribute

pidarped committed:
Make CameraAVStreamMgmt handle list attributes inside cluster server implementation.

Add Getter functions for cluster attributes.

pidarped committed:
Add Delegate functions for loading pre-allocated streams.

Add persistence support for the ranked video stream priorities
attribute.

pidarped committed:
Doc updates and adjustments based on latest Spec changes.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+    writer.Init(buffer);
+    ReturnErrorOnFailure(viewport.Encode(writer, TLV::AnonymousTag()));
+
+    return mPersistentStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::CameraAVStreamMgmtViewport().KeyName(), buffer,
+                                               static_cast<uint16_t>(writer.GetLengthWritten()));
```
I do wonder: is there a reason to not use the key format that the other attributes (via SafeAttributePersistenceProvider) use?

And the reason I ask that is this: as soon as you have two instances of this cluster on different endpoints, this setup breaks completely, with the global key name you have here.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+CHIP_ERROR CameraAVStreamMgmtServer::ClearViewport()
+{
+    // TODO
```
Needs to be at least tracked in an issue that is referenced here.

pidarped replied:
Removed TODO and added the call to delete key.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+    writer.Init(buffer);
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
```
This, and the way viewport is stored, assumes there will never be format/versioning changes, right?  Is that a good assumption?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+        DefaultStorageKeyAllocator::CameraAVStreamMgmtRankedVideoStreamPriorities().KeyName(), buffer, len));
+
+    TLV::TLVReader reader;
+    reader.Init(buffer);
```
Shouldn't this somehow take `len` into account?  The viewport codepath did this right.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+        ReturnErrorOnFailure(reader.Get(mRankedVideoStreamPriorities[i]));
+    }
+
+    ReturnErrorOnFailure(reader.ExitContainer(arrayType));
```
Whoa, wait.  If our Next() failed because the TLV was malformed, we need to return at that point.  Next() is only allowed to return "end of TLV" or success for us to keep going.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+    }
+    else
+    {
+        VerifyOrExit(!commandData.watermarkEnabled.HasValue(), status = Status::InvalidCommand;
```
I strongly suggest not using "go to exit".  Have this function return a status and have the caller call AddStatus on failure?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+    if (HasFeature(Feature::kWatermark))
+    {
+        isWaterMarkEnabled = commandData.watermarkEnabled.ValueOr(false);
```
This allows invalid commands where we have the WMARK feature but the command is missing the field.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+    if (HasFeature(Feature::kOnScreenDisplay))
+    {
+        isOSDEnabled = commandData.OSDEnabled.ValueOr(false);
```
Again, this allows invalid commands.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+                     ChipLogError(Zcl, "CameraAVStreamMgmt: Failed to allocate Video stream. Unsupported OnScreenDisplay feature"));
+    }
+
+    // Call the delegate
```
What happened to verifying that the passed-in enums have valid valies?  That the min/max constraints of all sorts are satisfied?  All of that needs to happen here, because relying on the delegate to do that checking means it won't happen.

pidarped replied:
Added sanity checks for args for all the commands.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+    if (status == Status::Success)
+    {
+        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
```
I don't get it.  Where did `response.videoStreamID` get set?

Was this tested at all?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+    if (HasFeature(Feature::kWatermark))
+    {
+        isWaterMarkEnabled = commandData.watermarkEnabled.ValueOr(false);
```
This has all the same bugs as VideoStreamAllocate in terms of allowing invalid commands, not validating enums, etc.

I'm stopping here in the .cpp.  Please fix the issues pointed out so far, including in the code I have not read yet.

pidarped committed:
Address review comments.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+CHIP_ERROR CameraAVStreamMgmtServer::SetRankedVideoStreamPriorities(const StreamUsageEnum newPriorities[kNumOfStreamUsageTypes])
+{
+    std::copy(newPriorities, newPriorities + kNumOfStreamUsageTypes, mRankedVideoStreamPriorities);
```
Replaced.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+    VideoStreamAllocate(StreamUsageEnum streamUsage, CameraAvStreamManagement::VideoCodecEnum videoCodec,
+                        const uint16_t minFrameRate, const uint16_t maxFrameRate, VideoResolutionStruct minResolution,
+                        VideoResolutionStruct maxResolution, const uint32_t minBitRate, const uint32_t maxBitRate,
+                        const uint16_t minFragmentLen, const uint16_t maxFragmentLen, bool waterMarkEnabled, bool osdEnabled) = 0;
```
I presume for this call you are talking about the WMARK and OSD features. The checks for the feature are done in the handler before calling this delegate function. If feature is not supported these would default to false.

bzbarsky-apple replied:
My question is why we would pass false to the delegate instead of correctly representing the "not present" state?

pidarped replied:
Converted to optional args and passed as-is to the delegate.

ksperling-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+     * appropriate error.
+     */
+    virtual Protocols::InteractionModel::Status
+    SetStreamPriorities(const DataModel::DecodableList<StreamUsageEnum> streamPriorities) = 0;
```
`Span<StreamUsageEnum>` ?

pidarped replied:
I realized that I had added the stream priorities as a member array. So, no need to pass to the delegate. Instead, there is a notification callback that is invoked on the delegate when the stream priorities is set via the Command.

ksperling-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,656 @@
+     *  respective streams.
+     */
+    virtual Protocols::InteractionModel::Status
+    LoadAllocatedVideoStreams(std::vector<VideoStreamStruct> & allocatedVideoStreams) = 0;
```
Has there been a conscious decision to use `std::vector` in this code? Most core code avoids using it.

pidarped replied:
I have used vectors here mainly for ease of use for managing dynamic lists, and especially, since cameras are expected to be relatively resourceful. What is the preferred approach to represent dynamic lists that are owned?

yufengwangca replied:
If the objects are very large/non-copyable, then we might use std::vector<std::unique_ptr<VideoStreamStruct>>?
 
std::unique_ptr clearly expresses unique ownership: the vector “owns” the objects. This helps avoid memory leaks and clarifies that nothing else should be responsible for those objects’ lifetimes.

pidarped replied:
Allocation should be infrequent and the number of streams should be relatively small. Keeping it as a vector of structs seems simpler to manage.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+constexpr size_t kNumOfStreamUsageTypes = 4;
+
+// StreamUsageEnum + Anonymous tag(1 byte)
+constexpr size_t kStreamUsageTlvSize = sizeof(StreamUsageEnum) + 1;
```
To avoid discrepancy, I am thinking of using a sizeof(uint8_t) for the StreamUsageEnum.

bzbarsky-apple replied:
No, that's not the point.  The point is that technically nothing requires a number that fits in uint8 to be encoded as one byte in the TLV.  Our TLV encoder does happen to do that, and if we are depending on that, we should probably fail out during encoding if that's violated.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+     * @param aEndpointId                       The endpoint on which this cluster exists. This must match the zap configuration.
+     * @param aClusterId                        The ID of the Camera AV Stream Management cluster to be instantiated.
+     * @param aFeature                          The bitmask value that identifies which features are supported by this instance.
+     * @param aOptionalAttrs                    The bitmask value that identifies the optional attributes supported by this
```
Because in the spec a "mapN" doesn't mean each bit is independent; some mapN in the spec use two bits to represent stuff.  So you can't use BitFlags for spec mapN when generating code, because you don't know the details of those semantics....

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+     * @param aEndpointId                       The endpoint on which this cluster exists. This must match the zap configuration.
+     * @param aClusterId                        The ID of the Camera AV Stream Management cluster to be instantiated.
+     * @param aFeature                          The bitmask value that identifies which features are supported by this instance.
+     * @param aOptionalAttrs                    The bitmask value that identifies the optional attributes supported by this
```
Why wouldn't we just fail out of Init on invalid input, especially since the API docs say we do?

pidarped replied:
Added optional bitflag checks along with the feature checks in the Init function.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,627 @@
+    CameraAVStreamMgmtDelegate & mDelegate;
+    EndpointId mEndpointId;
+    ClusterId mClusterId;
+    BitMask<Feature> mFeature;
```
Have made the BitFlags to be const.

pidarped committed:
Make watermark and OSD optional args.

pidarped committed:
Fix

pidarped committed:
Add feature and optional flags check in Init

pidarped committed:
Modify SetStreamPriorities

pidarped committed:
Add error handling for command params.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1801 @@
+
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(CommandHandlerInterfaceRegistry::Instance().RegisterCommandHandler(this));
+    return CHIP_NO_ERROR;
```
Added the feature and optional attribute checks.

pidarped committed:
Add CameraAvStreamMgmt cluster partially inside all-clusters-app.zap config

Make the snapshot attributes and commands of the CameraStreamMgmt
cluster be part of the all-clusters-app(in a minimalistic way) to
exercise the cluster.

yunhanw-google reviewed (commented):
```
2025-02-12 19:21:41 fcd452e3b655 PythonMatterControllerTEST[4749] ERROR Unexpected error occurred when running tests: Cannot decode value for path 2/<class \'chip.clusters.Objects.CameraAvStreamManagement.Attributes.SupportedSnapshotParams\'>, got error: \'InteractionModelError: Failure (0x1)\', raw TLV data: \'None\'
2025-02-12 19:21:41 fcd452e3b655 PythonMatterControllerTEST[4749] ERROR Cannot decode value for path 2/<class \'chip.clusters.Objects.CameraAvStreamManagement.Attributes.SupportedSnapshotParams\'>, got error: \'InteractionModelError: Failure (0x1)\', raw TLV data: \'None\'
Traceback...
```

From cirque mobile test, we see the above error, it looks there is decoding error for CameraAvStreamManagement.Attributes.SupportedSnapshotParam

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+
+    // Call the delegate
+    status =
+        mDelegate.VideoStreamAllocate(streamUsage, videoCodec, minFrameRate, maxFrameRate, minResolution, maxResolution, minBitRate,
```
Where is the constraint validation for fields of minResolution/maxResolution?

pidarped committed:
Add more constraint checking for command params.

Add a delegate notification callback for attribute changes.
Address other review comments.

pidarped committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

pidarped committed:
Restyle fixes.

pidarped committed:
Add CameraAVStreamMgmt cluster to EP1 of camera-app.zap.

pidarped committed:
Remove Enum validation functions.

pidarped committed:
Address review comments

- Add endpoint id to logs.
- Add missing constraint checks.

pidarped committed:
Various Fixes

- Remove usage of DefaultStorageAllocator for the TLV encoded stored
  attributes, and replace them with SafeWriteValue()/SafeReadValue()
  APIs from GetSafeAttributePersistenceProvider.
- Make the RankedStreamPriorities a vector and handle the TLV encoding
  of the list of enums.
- Populate the CaptureSnapshot response properly with image data
  received from delegate.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+    ReturnErrorOnFailure(viewport.Encode(writer, TLV::AnonymousTag()));
+
+    StorageKeyName key =
+        DefaultStorageKeyAllocator::AttributeValue(mEndpointId, CameraAvStreamManagement::Id, Attributes::Viewport::Id);
```
Didn't see the SafeWriteValue/SafeReadValue() APIs earlier for writing/reading TLV encoded bytespans. Have changed to that. Thanks.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+    return mPersistentStorage->SyncSetKeyValue(key.KeyName(), buffer, static_cast<uint16_t>(writer.GetLengthWritten()));
+}
+
+CHIP_ERROR CameraAVStreamMgmtServer::ClearViewport()
```
I initially thought I would clear before storing but it seems it is not required. Removed it.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+
+    for (uint8_t i = 0; i < kNumOfStreamUsageTypes; i++)
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_UnsignedInteger, TLV::AnonymousTag()));
```
Have changed the attribute to be a vector. There are some changes that are coming in the Spec where the device would have a SupportedStreamUsages attribute and we will check against that when processing this command. So, changing it to a vector now, and will be updating with those checks when the spec changes land.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+    int i = 0;
+    while (iter.Next())
+    {
+        auto & streamUsage          = iter.GetValue();
```
Added check.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+    while (iter.Next())
+    {
+        auto & streamUsage          = iter.GetValue();
+        rankedStreamPriorities[i++] = streamUsage;
```
Changed to a vector and not a fixed sized array.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+        rankedStreamPriorities[i++] = streamUsage;
+    }
+
+    VerifyOrReturn(i == kNumOfStreamUsageTypes, {
```
Changed to a vector for storing the StreamUsageEnums, which can have less number of items than kNumOfStreamUsageTypes. When the new Spec attributed of SupportedStreamUsages lands, we would need to add checks against that when this command is processed.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1765 @@
+{
+
+    auto & streamPriorities = commandData.streamPriorities;
+    // auto & streamPriorities = commandData.streamPriorities;
```
Removed spurious commented line.

pidarped committed:
Add parameters struct for housing the set of stream allocation parameters for the delegate.

pidarped committed:
Add initial delegate implementation in all-clusters-app.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        break;
+    case SpeakerMuted::Id:
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just check Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        break;
+    case SpeakerVolumeLevel::Id:
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just check Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        break;
+    case SpeakerMaxLevel::Id:
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just check Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+            ChipLogError(Zcl, "CameraAVStreamMgmt[ep=%d]: can not get SpeakerMaxLevel, feature is not supported", mEndpointId));
+        ReturnErrorOnFailure(aEncoder.Encode(mSpeakerMaxLevel));
+        break;
+    case SpeakerMinLevel::Id:
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just check Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+    switch (aPath.mAttributeId)
+    {
+    case HDRModeEnabled::Id: {
+        // Optional Attribute if Video is supported
```
Not in the spec PR this is claiming to implement.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        ReturnErrorOnFailure(aDecoder.Decode(viewPort));
+        return SetViewport(viewPort);
+    }
+    case SpeakerMuted::Id: {
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        ReturnErrorOnFailure(aDecoder.Decode(speakerMuted));
+        return SetSpeakerMuted(speakerMuted);
+    }
+    case SpeakerVolumeLevel::Id: {
+        VerifyOrReturnError(
+            HasFeature(Feature::kAudio) && HasFeature(Feature::kSpeaker), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
```
Just Speaker.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,662 @@
+// The number of possible values of StreamUsageEnum.
+constexpr size_t kNumOfStreamUsageTypes = 4;
+
+// StreamUsageEnum + Anonymous tag ( 1 byte ).
+// Assumes min-size encoding ( 1 byte ) for the integer.
```
```suggestion
// StreamUsageEnum + Anonymous tag (1 byte).
// Assumes min-size encoding (1 byte) for the integer.
```

"parentheticals" and "parentheses" are not the same thing.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,662 @@
+    struct VideoStreamAllocateArgs
+    {
+        StreamUsageEnum
+            streamUsage; // Indicates the type of usage of stream ( Recording, Liveview, etc ) that this allocation is for.
```
Look, things with parentheses should be like this: "abc (def ghi) something".  Space before open and after close (which you used to not have) and no spaces after open and before close (which you added for some reason all over here).

Please make this stuff look sane.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+    ReturnErrorOnFailure(viewport.Encode(writer, TLV::AnonymousTag()));
+
+    auto path = ConcreteAttributePath(mEndpointId, CameraAvStreamManagement::Id, Attributes::Viewport::Id);
+    return GetSafeAttributePersistenceProvider()->SafeWriteValue(path, ByteSpan(buffer));
```
This is wrong.  It should be creating a span with the right length.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+
+    auto path = ConcreteAttributePath(mEndpointId, CameraAvStreamManagement::Id, Attributes::Viewport::Id);
+    ReturnErrorOnFailure(GetSafeAttributePersistenceProvider()->SafeReadValue(path, bufferSpan));
+    bufferSpan.reduce_size(size);
```
This is _very very_ wrong.  Please fix.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+
+    auto path = ConcreteAttributePath(mEndpointId, CameraAvStreamManagement::Id, Attributes::RankedVideoStreamPrioritiesList::Id);
+    ReturnErrorOnFailure(GetSafeAttributePersistenceProvider()->SafeReadValue(path, bufferSpan));
+    bufferSpan.reduce_size(size);
```
This is wrong.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        i++;
+    }
+
+    VerifyOrReturnError(err == CHIP_END_OF_TLV, err);
```
OK, so here if the stored TLV had more than kNumOfStreamUsageTypes entries, we will return CHIP_NO_ERROR at this point, right, without doing the ExitContainer/VerifyEndOfContainer below.  Is that the expected behavior?

pidarped replied:
kNumOfStreamUsageTypes is expected to hold the max number of supported stream usages. So, the stored list is not expected to have more than kNumOfStreamUsageTypes entries. However, we have added a new attribute `SupportedStreamUsages` and the `RankedStreamPriorities` list should conform to that. I have filed an issue(https://github.com/project-chip/connectedhomeip/issues/37933) for adding that logic.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+    auto & videoStreamID      = commandData.videoStreamID;
+
+    // VideoStreamModify should have either the WMARK or OSD feature supported
+    VerifyOrReturn(HasFeature(Feature::kWatermark) || HasFeature(Feature::kOnScreenDisplay),
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand));
```
Why is this checked here and not in the dispatch where Video feature is checked?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+    });
+
+    VerifyOrReturn(commandData.bitRate > 0, {
```
Where are the constraint checks on minResolution/maxResolution?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        }
+        rankedStreamPriorities.push_back(streamUsage);
+    }
+
```
Again: Where's the success check for the list iteration?

As in, https://github.com/project-chip/connectedhomeip/pull/36204#discussion_r1958725972 was not addressed.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+    auto & snapshotStreamID    = commandData.snapshotStreamID;
+    auto & requestedResolution = commandData.requestedResolution;
+    ImageSnapshot image;
+
```
Where is the constraint validation for requestedResolution?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+        return;
+    }
+
+    if (image.data.size() > kMaxSnapshotImageSize)
```
Where is kMaxSnapshotImageSize coming from?  I don't see it in the spec.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1850 @@
+    }
+
+    // copy the image data into the response
+    memcpy(&response.data, image.data.data(), image.data.size());
```
You just smashed the stack.  This is a critical security bug right here.  Please let me know if you can't figure out why and how to do this part right.

pidarped replied:
Presumed the response.data to be a byte array. Changed to a ByteSpan.

pidarped commented:
Trying to incorporate changes in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11309 which was in flux. Not all changes from that spec PR have been included yet. Will indicate here once it is ready for review again.

pidarped committed:
Update Camera AVStreamMgmt Server SDK based on Spec changes from Spec PR#11309.

pidarped committed:
Update all-clusters-app and camera-app with attribute changes.

pidarped committed:
Updates to CameraAVStreamMgmt delegate.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,662 @@
+
+    virtual ~CameraAVStreamMgmtDelegate() = default;
+
+    struct VideoStreamAllocateArgs
```
The VideoStreamStruct is similar with a ref count field. Replaced with that.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,662 @@
+        chip::Optional<bool> isOSDEnabled;       // Indicates  whether the on-screen display can be applied on the video stream.
+    };
+
+    struct AudioStreamAllocateArgs
```
Replaced with AudioStreamStruct::Type.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,662 @@
+        uint8_t bitDepth;          // Indicates the number of information bits ( 8, 16, 24 or 32 ) used to represent each sample.
+    };
+
+    struct SnapshotStreamAllocateArgs
```
Replaced with SnapshotStreamStruct::type

pidarped committed:
Fix Command de-allocation comments.

pidarped commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    return Status::Success;
```
Updated comment.

pidarped commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    return Status::Success;
```
Updated comment.

pidarped commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    return Status::Success;
```
Updated comment.

pidarped committed:
Fix Command code numbering in zap file.

pidarped committed:
Add SnapshotStreamModify command support.

pidarped committed:
Fix persisted storage handling for Viewport and STream Priorities.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,636 @@
+    /**
+     *   @brief Handle Command Delegate for Video stream allocation with the provided parameter list.
+     *
+     *   @param[in]  allocateArgs   Structure with parameters for video stream allocation.
```
Should document that callee must ignore the videoStreamID and referenceCount values?  Similar for the other Allocate methods.

pidarped committed:
Address review comments.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -0,0 +1,1908 @@
+        i++;
+    }
+
+    ReturnErrorOnFailure(reader.ExitContainer(arrayType));
```
Yes, added the `VerifyOrReturnError(err == CHIP_ERROR_END_OF_TLV)`

pidarped committed:
Resolve conflicts in all-clusters-app.matter and camera-app.matter.

bzbarsky-apple commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,373 @@
+    {
+        if (!stream.isAllocated)
+        {
+            foundAvailableStream = true;
```
This works, but why do we need a separate loop?  Why can't we have one loop which does:
```
        if (!stream.isAllocated)
        {
            foundAvailableStream = true;
            if (stream.codec == allocateArgs.videoCodec)
            {
                stream.isAllocated = true;
                outStreamID        = stream.id;
                return Status::Success;
            }
        }
```
kind of thing?

bzbarsky-apple commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,373 @@
+        }
+    }
+
+    return Status::Failure;
```
Still needs a spec issue and comment pointing to it here that the behavior on codec mismatch is not defined.  FAILURE is not a good error code to be returning in general.

pidarped replied:
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/11384

bzbarsky-apple commented on examples/all-clusters-app/all-clusters-common/src/camera-av-stream-delegate-impl.cpp:
```diff
@@ -0,0 +1,373 @@
+
+    bool foundAvailableStream = false;
+
+    for (AudioStream & stream : audioStreams)
```
See above.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,644 @@
+    /**
+     *   @brief Handle Command Delegate for Video stream allocation with the provided parameter list.
+     *
+     *   @note The videoStreamID and referenceCount fields in the struct are
```
```suggestion
     *   @note The videoStreamID and referenceCount fields in the struct must be
```

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,644 @@
+    /**
+     *   @brief Handle Command Delegate for Video stream allocation with the provided parameter list.
+     *
+     *   @note The videoStreamID and referenceCount fields in the struct are
```
Shouldn't this be part of the param definition?  Or after it and say "fields in allocateArgs"?  Because the obvious question here is "what struct"?

pidarped replied:
Moved it under allocateArgs.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,644 @@
+    /**
+     *   @brief Handle Command Delegate for Audio stream allocation.
+     *
+     *   @note The audioStreamID and referenceCount fields in the struct are
```
See above.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -0,0 +1,644 @@
+    /**
+     *   @brief Handle Command Delegate for Snapshot stream allocation.
+     *
+     *   @note The snapshotStreamID and referenceCount fields in the struct are
```
See above.

pidarped committed:
Update src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

pidarped committed:
Address review comments.
